### ISSUE TITLE
Allow multiple share links

### ISSUE THREAD

**Issue by schiessle** [nextcloud/server#11169]
Before starting on the implementation of multiple share links I would like to get some feedback from @nextcloud/designers  how this should look like.

I already thought about it a little bit and this is my suggestion:

1. Initial share view:

![1](https://user-images.githubusercontent.com/1589737/45354055-0ad55680-b5bd-11e8-8afa-b773e9d33b49.png)

If you click on "create new public link" you get a text input field which is pre-filled with "share link 1", the next one "share link 2", and so on, so that the user can give the link a name.

![2](https://user-images.githubusercontent.com/1589737/45354150-4cfe9800-b5bd-11e8-8b11-57dcc0e7781c.png)

If the link was created it will look like this, the user sees the new share link, with the '...' menu and again a option to create another public link:

![3](https://user-images.githubusercontent.com/1589737/45354162-57209680-b5bd-11e8-929e-e3ab279b95b3.png)

To the '...' menu I would add the two new additional option "rename" and "unshare".

Opionions? cc @jancborchardt

# Steps to complete it

- [ ] agree on design
- [ ] front-end implementation
- [ ] back-end implementation
- [ ] OCS API implementation
- [ ] Tests

**Comment by MariusBluem** [comment]
See discussion and mockups here: 
https://github.com/nextcloud/server/issues/5336

**Comment by schiessle** [comment]
> When you tap into the share input field, "Add new share link" should be suggested as entry like in @MariusBluem's mockup in the linked issue.

And the link, after it was created, looks then the same way as the first one (with the checkbox and so on). Do they all have the same name? Or do you just add a number? How will people distinguish the different links?

Also on link removal. Unchecking the checkbox will then remove the complete entry as long as there is one more link and if it is the last one we keep the entry and just uncheck it so that the user can again create the first public link.

Not sure if I understand it completely but sounds quite clumsy and more complicated then what I suggested.

Btw, your mockup here looks quite similiar to what I suggested: https://github.com/nextcloud/server/issues/5336#issuecomment-363023687 You also have always a "new share link" action on the top and below the existing shares, including already created public links.

**Comment by MariusBluem** [comment]
You should definitely be able to give share links a name (maybe we should make this mandatory at all ...)

**Comment by tobiasKaminsky** [comment]
> maybe we should make this mandatory at all 

I am unsure how many sharing links are done on one file, but if you only have one, there is no need to have a name for it (of course you should be able to name it, if you want).

**Comment by schiessle** [comment]
> You should definitely be able to give share links a name (maybe we should make this mandatory at all ...)

I agree, that's why my mockup gives you directly a input field, where you can just press enter or focus out to accept the default or give it a custom name without a lot of hassle.

**Comment by MorrisJobke** [comment]
> Is this a feature that should also be available on clients?

Would make sense 👍

**Comment by blackcrack** [comment]
would be nice if to be able for make a "public" folder to share it 
and give a frame from the NextCloud /Public-Files
where listing like this : 
http://www.blackysgate.de/files/folder-root/

to have a public-sharing like : 
https://webserver.org/nextcloud/User/public-files

therewith it is possible for provide directories and files in public for download,
add in a downloadpage or in a webpage to have a possible for publishing files and folders in easy way..

therewith it is also possible to put videos on this public-sharing folders 
and if we have a webm(av1) ( [ [request] Webm Encoder app for Files-app ](https://help.nextcloud.com/t/request-webm-encoder-app-for-files-app/38961) )

best regards
Blacky

**Comment by tobiasKaminsky** [comment]
> > Is this a feature that should also be available on clients?
> 
> Would make sense

Once there is a PR, I'll open up an issue on all clients (prior does not make sense, as we cannot implement/test it)
For UI/UX on clients I assume, that we can try to stick very close to webUI.

**Comment by jancborchardt** [comment]
@schiessle the main issue with the flow you suggest is that people are faced with an input field first. This is a barrier – you want to simply share a link, not think of a name. There’s a reason we designed it like we did. ;) Also because the most common case is only having 1 single share link, which in your design would take up 2 rows already – confusing people.

> And the link, after it was created, looks then the same way as the first one (with the checkbox and so on). Do they all have the same name? Or do you just add a number? How will people distinguish the different links?

No, it could simply be "Share link 2" for example. Or even better, name it like the settings it uses "Read-only link", "Files drop link", etc. The default is "Share link" and is editable. You need a fallback for that too even when using an input field.

Also DO NOT use the wording "Public link" as the link is not publicized anywhere.

-----

TL;DR: There is a design we discussed among the design team, I described it above and it’s in the linked issue https://github.com/nextcloud/server/issues/5336. Let’s please go with what we agreed on as a group, ok? :slightly_smiling_face:

**Comment by jancborchardt** [comment]
Yes, we need to skip the entry field. And yes, we should also not disrupt the current flow. And yes, when you only have a single shared link, there should not be 2 entries in the links related to link sharing.

There should be one default entry in the list for "Share link" exactly like now.

And additional links you should be able to create by:
- opening the 3-dot menu of the default share link entry and having a "Create another share link" entry there
- tapping in the input field and having a "Create another share link" entry there

(If you only think one is good, decide for one.)

Then that entry appears and the link is immediately copied into the clipboard. And the 3-dot menu is opened as well. (As per @skjnldsv pull request.)

Ok?

**Comment by skjnldsv** [comment]
@schiessle I'm on it until the feature freeze, this is the only task left for me :)
Status: this is a nightmare. The code is a complete mess and I'm banging my head.

I'm very disappointed that we did not wait for the new sidebar standard so we could at least know the code we're working on :'(

**Comment by schiessle** [comment]
thanks a lot @skjnldsv ! So it is at least not only me who felt lost in the code...

**Comment by skjnldsv** [comment]
@schiessle unfortunately no. We'll have to talk about this at the next hackweek :)

**Comment by skjnldsv** [comment]
@jancborchardt can you start a review now?
That would increase my response time to fixes to day for the feature freeze :)

Basically the only thing left is the label handling! How should we do that?
1. Add an edit button?
2. In the popovermenu?
